packages/components/src/DataList/DataList.utils.tsx

```diff
@@ -137,3 +165,152 @@ export function generateDataListEmptyState({
+    layouts,
+    (layout: React.ReactElement<DataListLayoutProps<T>>) => {
+      return elementData.map((child, i) => {
+        // TODO: Don't use index as key. Might have to force an ID on the data JOB-76773
+        return (
+          <div className={styles.listItem} key={`${i}-${child.id}`}>
+            {layout.props.children(child)}
+          </div>
+        );
+      });
+    },
+    mediaMatches,
+  );
+}
+
+export function renderDataListHeader<T extends DataListObject>(
+  layouts: React.ReactElement<DataListLayoutProps<T>>[] | undefined,
+  headerData?: DataListItemTypeFromHeader<T, DataListHeader<T>>,
+  mediaMatches?: Record<Breakpoints, boolean>,
+) {
+  return renderDataListLayouts(
+    layouts,
+    (layout: React.ReactElement<DataListLayoutProps<T>>) => {
+      const showHeader = layout.props.showHeader ?? true;
+      return (
+        showHeader &&
+        headerData && (
+          <div className={styles.header}>
+            {layout.props.children(headerData)}
+          </div>
+        )
+      );
+    },
+    mediaMatches,
+  );
+}
+
+function renderDataListLayouts<T extends DataListObject>(
+  layouts: React.ReactElement<DataListLayoutProps<T>>[] | undefined,
+  renderLayout: (
+    layout: React.ReactElement<DataListLayoutProps<T>>,
+  ) => ReactNode,
+  mediaMatches?: Record<Breakpoints, boolean>,
+) {
+  const sizePropsOfLayouts = layouts?.map(layout => layout.props.size || "xs");
+  return layouts?.map(layout => {
+    const layoutSize = layout.props.size || "xs";
+
+    const isVisible = isLayoutVisible({
+      layoutSize,
+      mediaMatches,
+      sizePropsOfLayouts,
+    });
+
+    return isVisible && renderLayout(layout);
+  });
+}
+
+/**
+ *Determine is layout is visible by checking a media query matches for the
+  visible sizes of the layout and there isn't a larger layout that should be rendered
+ */
+function isLayoutVisible({
+  layoutSize,
+  mediaMatches,
+  sizePropsOfLayouts,
+}: {
+  layoutSize: Breakpoints;
+  mediaMatches?: Record<Breakpoints, boolean>;
+  sizePropsOfLayouts?: Breakpoints[];
+}) {
+  const largerBreakpointsToHide = sizePropsOfLayouts?.filter(
+    size => BREAKPOINTS.indexOf(size) > BREAKPOINTS.indexOf(layoutSize),
+  );
+  const sortedLargerBreakpoints = sortSizeProp(largerBreakpointsToHide || []);
+
+  const visibleBreakpoints = BREAKPOINTS.slice(
+    BREAKPOINTS.indexOf(layoutSize),
+    sortedLargerBreakpoints[0]
+      ? BREAKPOINTS.indexOf(sortedLargerBreakpoints[0])
+      : undefined,
+  );
+  return (
+    visibleBreakpoints.some(breakpoint => {
+      return Boolean(mediaMatches?.[breakpoint]);
+    }) &&
+    largerBreakpointsToHide?.reduce<boolean>((acc, breakpoint) => {
+      return acc && !mediaMatches?.[breakpoint];
+    }, true)
+  );
+}
+
+export function sortSizeProp(sizeProp: Breakpoints[]) {
+  return sizeProp.sort(
+    (a, b) => BREAKPOINTS.indexOf(a) - BREAKPOINTS.indexOf(b),
+  );
+}
+
+export function useGridLayoutMediaQueries() {
```

Comment:
Chose this approach because `useResizeOberserver` would cause the entire table to rerender a bunch as the window resizes as opposed to this approach which only rerenders when a mediaQuery matches

Reply:
We could add something like this to our hooks package but it is far outside the scope of this subtask

Reply:
agreed

Reply:
Actually, I wonder if this should be in its own hook file rather than with the utils? Also, just curious, why is it called grid?